src/treq/_utils.py

```diff
@@ -45,3 +48,50 @@ def default_pool(reactor, pool, persistent):
         set_global_pool(HTTPConnectionPool(reactor, persistent=True))
 
     return get_global_pool()
+
+
+def warn(*args, **kwargs):
+    """
+    Utility function to raise warnings with stack level set automatically to
+    the first position outside a given namespace.
+
+    Notice that if you pass the stacklevel keyword, it will be increased by 1
+    in order to ignore this wrapper function and that will be passed directly
+    to warnings.warn. That is: it is functionally equivalent to warnings.warn.
+
+    @param namespace: calls within this namespace (aka startswith) will not
+        be accounted for in the stacklevel variable if possible.
+        If not defined, the default is the top level package of this code.
+    @param default_stacklevel: in those cases in which there is no code
+        outside of namespace in the stack, we will default to this value.
+        This utility function already takes into account the fact that it is a
+        wrapper around warnings.warn and secretly passes default_stacklevel+1
+        when necessary.
+        Defaults to 1, which would match the place where this was called from.
+    """
+    # Default to top-level package
+    namespace = kwargs.pop('namespace', __name__.partition('.')[0])
```

Comment:
The default could be gotten by grabbing the calling stack frame's module's name, if it has one.  We're already inspecting like we just don't care, why not a bit more.

Reply:
In so far as it needs to support other packages using it, perhaps it should get a class.  Any code that uses it can create an instance which holds the desired top level package name and then call that instance instead of having to reinspect or repass the top level package.  With it only needing to be passed once it's not a hassle to anyone to just require it be specified and never guess.  Then we can disregard my last comment and have a tad less inspection.